src/Nethermind/Nethermind.State/Witnesses/WitnessCollector.cs

```diff
@@ -0,0 +1,122 @@
+//  Copyright (c) 2018 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+using Nethermind.Trie;
+
+namespace Nethermind.State.Witnesses
+{
+    /// <summary>
+    /// <threadsafety static="true" instance="false" />
+    /// </summary>
+    public class WitnessCollector : IWitnessCollector
+    {
+        private LruCache<Keccak, IReadOnlyList<Keccak>> _witnessCache
+            = new LruCache<Keccak, IReadOnlyList<Keccak>>(256, "Witnesses");
+        
+        public IReadOnlyCollection<Keccak> Collected => _collected;
+
+        public WitnessCollector(IKeyValueStore? keyValueStore, ILogManager? logManager)
+        {
+            _keyValueStore = keyValueStore ?? throw new ArgumentNullException(nameof(keyValueStore));
+            _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        }
+
+        public void Add(Keccak hash)
+        {
+            _collected.Add(hash);
+        }
```

Comment:
_collected is HashedSet, what if we add/touch same hash twice? It will be in order of first Touch not second.
Also HashSet doesn't guarantee order, but should keep it when there are only insertions I think. If order is important we need to at least comment on it.

Reply:
order does not matter